client/dfget/dfget.go

```diff
@@ -20,160 +20,219 @@ import (
+		} else {
+			var (
+				start      = time.Now()
+				stream     *daemonclient.DownResultStream
+				result     *dfdaemon.DownResult
+				pb         *progressbar.ProgressBar
+				request    = newDownRequest(cfg, hdr)
+				downLength uint64
+			)
+
+			if stream, downError = client.Download(ctx, request); downError == nil {
+				if cfg.ShowProgress {
+					pb = newProgressBar(-1)
+				}
+
+				for {
+					if result, downError = stream.Recv(); downError != nil {
+						break
+					}
+
+					if result.CompletedLength > 0 {
+						downLength = result.CompletedLength
+						if pb != nil {
+							_ = pb.Set64(int64(downLength))
+						}
+					}
+
+					// success
+					if result.Done {
+						if pb != nil {
+							pb.Describe("Downloaded")
+							_ = pb.Close()
+						}
+
+						wLog.Infof("download from daemon success, length:%dByte cost:%dms", downLength, time.Now().Sub(start).Milliseconds())
+						fmt.Printf("finish total length %d Byte\n", downLength)
+
+						break
+					}
+				}
+			}
+
+			if downError != nil {
+				wLog.Warnf("daemon downloads file error:%v", downError)
+				fmt.Printf("daemon downloads file error:%v\n", downError)
+
+				downError = downloadFromSource(ctx, cfg, hdr)
 			}
-			break
-		}
-		if result.CompletedLength > 0 {
-			pb.Set64(int64(result.CompletedLength))
-		}
-		if result.Done {
-			pb.Describe("Downloaded")
-			pb.Finish()
-			end = time.Now()
-			fmt.Printf("Task: %s\nPeer: %s\n", result.TaskId, result.PeerId)
-			fmt.Printf("Download success, time cost: %dms, length: %d\n", end.Sub(start).Milliseconds(), result.CompletedLength)
-			break
 		}
+	}()
+
+	<-ctx.Done()
+
+	if ctx.Err() == context.DeadlineExceeded {
+		return errors.Errorf("download timeout(%s)", cfg.Timeout)
+	} else {
+		return downError
 	}
-	if err != nil {
-		logger.Errorf("download by dragonfly error: %s", err)
-		return downloadFromSource(cfg, hdr)
-	}
-	return nil
 }
 
-func downloadFromSource(cfg *config.DfgetConfig, hdr map[string]string) (err error) {
+func downloadFromSource(ctx context.Context, cfg *config.DfgetConfig, hdr map[string]string) error {
 	if cfg.DisableBackSource {
-		err = fmt.Errorf("dfget download error, and back source disabled")
-		logger.Warnf("%s", err)
-		return err
+		return errors.New("try to download from source but back source is disabled")
 	}
-	fmt.Println("dfget download error, try to download from source")
 
 	var (
+		wLog     = logger.With("url", cfg.URL)
 		start    = time.Now()
-		end      time.Time
 		target   *os.File
 		response io.ReadCloser
+		err      error
 		written  int64
 	)
 
-	response, err = source.Download(context.Background(), cfg.URL, hdr)
-	if err != nil {
-		logger.Errorf("download from source error: %s", err)
+	wLog.Info("try to download from source and ignore rate limit")
+	fmt.Println("try to download from source and ignore rate limit")
+
+	dir := filepath.Dir(cfg.Output)
+	_ = fileutils.MkdirAllWithOwner(dir, basic.UserID, basic.UserGroup)
```

Comment:
If the user did not has permission to create directories ?

Reply:
root setuid has permission to create